test: Orderbook TXE tests

benesjan opened the pull request:
Fixes #14525 

We didn't have the Orderbook thoroughly tested as when I implemented it TXE was not yet reasonable. Now it's reasonable so I decided to tackle this. Since AI is known to be good at tests I decided to gave it a try.

## AI experiment
This was the prompt + well curated the context window by opening the relevant tabs in Cursor:

> write the orderbook tests in noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr
> 
> note that the Orderbook contract accepts 2 token contracts on the input so you will need to deploy those first. Use the setup and setup and mint functionality from noir-projects/noir-contracts/contracts/app/token_contract/src/test/utils.nr.
> 
> From the orderbook you want to test the setup order and create order functions. There is also already a yarn-project/end-to-end/src/e2e_orderbook.test.ts test but that is written in TypeScript and noit Noir and it tests only the happy path. In the Noir tests I would also like to test unhappy paths. See tests in noir-projects/noir-contracts/contracts/app/token_contract/src/test/transfer_in_private.nr for inspiration on how to structure the Noir tests.

Overall it managed to implement it all. The only thing I need to help it with is that it didn't understand that it needs to refer the `token_contract` crate when deploying:

```diff
env.deploy("Token")
env.deploy("@token_contract/Token")
```

Would say this was caused by us not having enough examples and me not putting that into a context window.

Other than this it managed to resolve on its own one other issue (tried to mint with an account with token minter permissions).

The only other thing I did is that I nuked quite a few tests it did because it was for example testing that a function call fails when it doesn't have token transfer authwit provided. Felt like this is sufficient to have only in the token tests.

### Update on AI
Code quality was not great and had to redo quite a lot of stuff. That could be improved though if we had a proper reference (plenty of the style issues were present in the rest of our codebase).

benesjan commented:
* **#16451** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16451?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16451?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+
+// Utility function to setup orderbook with two tokens
+unconstrained fn setup_orderbook_with_tokens(
+    with_account_contracts: bool,
```
Given that the only contract tests that exist are are the ones we wrote, I think it'd be good for these to serve as a sort of reference that can be copied by others (e.g. like you did here). Because of that I'd focus not just on coverage but also overall style etc.

I don't like passing this boolean here for a number of reasons. One, it is a bit obsure (`setup(true)`), two, the owner doesn't really need to be an account contract since all they'll do is mint, not sign authwits.

benesjan replied:
Great point. The bool is ugly.

In this case I solved it by just moving the `minter` account creation out of the function. It made sense to do because the setup is only 1 line and we've already setup the `taker` account out of the `setup` in the `full_flow` test.

If you like how it looks now I can also update other tests were we have this ugly bool (Token and NFT tests). As you say seems to make sense to do as AI is able to do a useful job if you give it a good reference.

Also just renamed the function to `setup` as the original name didn't describe it well enough (it also sets up test env and the minter account).

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+        env.create_light_account()
+    };
+
+    let (token0_address, token1_address) = setup_tokens(&mut env, owner);
```
This account should not be called the 'owner', as there is no such thing. They are the token's minter.

benesjan replied:
Addressed 👍

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+    let orderbook_address =
+        env.deploy("Orderbook").with_public_initializer(owner, orderbook_initializer);
+
+    (env, orderbook_address, token0_address, token1_address, owner, owner)
```
We're returning the same account twice for two different purposes. During deployment we need to create a minter account, yes, but not a sender or recipient (like many other tests do), which I imagine this is.

nventuro replied:
After looking at this some more I noticed this last return value is never used. If using tools to craft these tests please make sure you review them yourself prior to submission.

benesjan replied:
Sorry. Should have been more thorough with checking this AI stuff. Nuked it. Thanks

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+    amount: u128,
+) {
+    env.call_private(owner, Token::at(token_address).mint_to_private(to, amount));
+}
```
This seems silly, why have this function? Just inline the call

benesjan replied:
Nuked the func

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+    assert_eq(
+        env.simulate_utility(Token::at(token_address).balance_of_private(address)),
+        expected_amount,
+    );
```
Same, these utility functions just hurt readability by creating unnecessary layers. The function just performs a single call. It'd be even shorter if inlined since the `token: Token` would be in scope and this would just be

```noir
    assert_eq(
        env.simulate_utility(token.balance_of_private(address)),
        expected_amount,
    );
```

benesjan replied:
Nuked

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+    let authwit_nonce = 1;
+
+    // Mint tokens to owner
+    mint_to_private(env, token0_address, owner, owner, bid_amount);
```
Would be better to create a separate account for the maker and mint for them. Here minter and maker are the same account.

benesjan replied:
Addressed 👍

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+
+#[test]
+unconstrained fn full_flow() {
+    let (mut env, orderbook_address, token0_address, token1_address, maker, _) =
```
admin and maker get mixed up here.

benesjan replied:
Addressed with the larger overall refactor.

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+    add_private_authwit_from_call_interface(owner, orderbook_address, transfer_call_interface);
+
+    // Create order
+    let _order_id = env.call_private(
```
If this is a `create_order` test, then it should test properties of the created order.

benesjan replied:
I decided to just nuke this test case as it's included in the `full_flow` one. Think that now it looks better.

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+            ask_amount,
+            authwit_nonce_create,
+        ),
+    );
```
Now this entire thing it might make sense to put into a `create_order` helper fn.

benesjan replied:
No longer relevant as I merged the create order and full flow tests.

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+    let bid_amount = 1000 as u128;
+    let ask_amount = 2000 as u128;
+    let authwit_nonce_create = 1;
+    let authwit_nonce_fulfill = 2;
```
You can use globals in uppercase for regular valid values that just get reused in multiple tests.

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr:
```diff
@@ -0,0 +1,331 @@
+            ask_amount,
+            authwit_nonce,
+        ),
+    );
```
```suggestion
    let _order_id = env.call_private(
        owner,
        Orderbook::at(orderbook_address).create_order(
            token0_address,
            token1_address,
            0,
            ASK_AMOUNT,
            CREATE_ORDER_AUTHWIT_NONCE,
        ),
    );
```

If using globals you can highlight how this is the exact same as the other tests except that value changed. You also end up with much less verbose tests, making reviewing easier and introducing less room for error. I'd also remove these (probably AI generated) comments of the form

```
#[test(fail with = "thing bad")]
fn test_fail_thing_bad() {
  // this test fails because thing bad
```

benesjan replied:
Addressed

benesjan commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test/utils.nr:
```diff
@@ -0,0 +1,44 @@
+use token::Token;
+
+// Sets up test env with orderbook with two tokens and a minter account.
+// TODO(#16560): Make it possible to return a contract instance directly from setup func
```
I feel like it might be worth tackling [this issue](https://github.com/AztecProtocol/aztec-packages/issues/16560) I just create soon to have a perfect reference. WDYT?

nventuro replied:
Yes, but at mentioned in that issue we can't to do this right now.

benesjan commented on yarn-project/end-to-end/src/e2e_orderbook.test.ts:
```diff
@@ -9,9 +9,7 @@ import { setup } from './fixtures/utils.js';
-// tests to TXE once TXE 2.0 is ready. Didn't write it in TXE now as there is no way to obtain public events and all of
-// TXE will be rewritten soon.
+// Unhappy path tests are written only in Noir.
 describe('Orderbook', () => {
```
Would you trust TXE enough to just nuke this tests now?

I would say not yet but curious about your opinion.

nventuro replied:
Hm. I'd say not yet because TXE does not yet run separate wallets for each of the accounts, and so it is a single PXE will all accounts involved But in this e2e we have the exact same setup I think.

Given that this is the only example we have of async partial note completion i'd keep it just for that. It excercises parts of the node api that we don't in other places.

nventuro replied:
(unless we had a test in which we tested that explicitly. in that case yea kill this e2e)

benesjan replied:
We don't have a TS test that would test async partial note completion. I only wrote that in a [noir test](https://github.com/AztecProtocol/aztec-packages/blob/41319fd3eaefd2cbceaae7c04df4f8b49beede15/noir-projects/noir-contracts/contracts/app/token_contract/src/test/transfer_to_private.nr#L24).

Added there a comment regarding why we keep that test around.

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test/test.nr:
```diff
@@ -0,0 +1,208 @@
+//
+// Note: Ideally this test would be split into 2 separate test cases - one for order creation and one for order
+// fulfillment, with the second test running on the state from the first test. Since this feature is not currently
+// supported, combining them into a single comprehensive test is the best approach.
```
What feature? Tests running off of the result of another one? I don't think we will ever want that tbh.

I see no reason not to have separate tests, they'll just run in parallel.

benesjan replied:
It just felt weird here that the setup of the remove liquidity test was basically exactly the add liquidity test. Like if there would be an issue in add liquidity the remove liquidity test would fail as well making running the second test off of the first one natural.

Anyway nuked the comment https://github.com/AztecProtocol/aztec-packages/pull/16451/commits/095e87a06e5ca9b0f95b4a2fc4ff4f9d95192160

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test/test.nr:
```diff
@@ -0,0 +1,208 @@
+unconstrained fn create_order_zero_bid_amount() {
+    let (mut env, orderbook_address, token0_address, token1_address, _minter) = setup();
+
+    // We create a light account for the maker (as opposed to a contract account) since we are not using authwits here.
```
No need to be so verbose and repetitive accross tests. This is already 20 lines long and is testing something trivial, we should try to have them be smaller when reasonable. I'd also inline the bid amount etc.

nventuro replied:
Same applies to other tests. You want to be able to read all of these tests, not fall alseep while you do it, not gloss over the details and actually spot what is going on in each.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/16451/commits/49d64ee44a940cf5d06966d561a39a8511c4b6b7

nventuro commented on noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test/test.nr:
```diff
@@ -0,0 +1,208 @@
+
+    let bid_amount = 0 as u128;
+
+    let _order_id = env.call_private(
```
```suggestion
    let _ = env.call_private(
```

Remove noise, if irrelevant it goes away. See comment above.

Commit:
test: Orderbook TXE tests

Fixes #14525

We didn't have the Orderbook thoroughly tested as when I implemented it TXE was not yet reasonable. Now it's reasonable so I decided to tackle this. Since AI is known to be good at tests I decided to gave it a try.

## AI experiment
This was the prompt + well curated the context window by opening the relevant tabs in Cursor:

> write the orderbook tests in noir-projects/noir-contracts/contracts/app/orderbook_contract/src/test.nr
>
> note that the Orderbook contract accepts 2 token contracts on the input so you will need to deploy those first. Use the setup and setup and mint functionality from noir-projects/noir-contracts/contracts/app/token_contract/src/test/utils.nr.
>
> From the orderbook you want to test the setup order and create order functions. There is also already a yarn-project/end-to-end/src/e2e_orderbook.test.ts test but that is written in TypeScript and noit Noir and it tests only the happy path. In the Noir tests I would also like to test unhappy paths. See tests in noir-projects/noir-contracts/contracts/app/token_contract/src/test/transfer_in_private.nr for inspiration on how to structure the Noir tests.

Overall it managed to implement it all. The only thing I need to help it with is that it didn't understand that it needs to refer the `token_contract` crate when deploying:

```diff
env.deploy("Token")
env.deploy("@token_contract/Token")
```

Would say this was caused by us not having enough examples and me not putting that into a context window.

Other than this it managed to resolve on its own one other issue (tried to mint with an account with token minter permissions).

The only other thing I did is that I nuked quite a few tests it did because it was for example testing that a function call fails when it doesn't have token transfer authwit provided. Felt like this is sufficient to have only in the token tests.

### Update on AI
Code quality was not great and had to redo quite a lot of stuff. That could be improved though if we had a proper reference (plenty of the style issues were present in the rest of our codebase).

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
Co-authored-by: benesjan <janbenes1234@gmail.com>
